br/pkg/lightning/backend/local/region_job.go

```diff
@@ -619,10 +666,169 @@ func (j *regionJob) fixIngestError(
+	*h = append(*h, x.(*regionJob))
+}
+
+func (h *regionJobRetryHeap) Pop() any {
+	old := *h
+	n := len(old)
+	x := old[n-1]
+	*h = old[0 : n-1]
+	return x
+}
+
+// regionJobRetryer is a concurrent-safe queue holding jobs that need to put
+// back later, and put back when the regionJob.waitUntil is reached. It maintains
+// a heap of jobs internally based on the regionJob.waitUntil field.
+type regionJobRetryer struct {
+	// lock acquiring order: protectedClosed > protectedQueue > protectedToPutBack
+	protectedClosed struct {
+		mu     sync.Mutex
+		closed bool
+	}
+	protectedQueue struct {
+		mu sync.Mutex
+		q  regionJobRetryHeap
+	}
+	protectedToPutBack struct {
+		mu        sync.Mutex
+		toPutBack *regionJob
+	}
+	putBackCh chan<- *regionJob
+	reload    chan struct{}
+	done      chan struct{}
+}
+
+func newRegionJobRetryer(putBackCh chan<- *regionJob) *regionJobRetryer {
+	ret := &regionJobRetryer{
+		putBackCh: putBackCh,
+		reload:    make(chan struct{}, 1),
+		done:      make(chan struct{}),
+	}
+	ret.protectedQueue.q = make(regionJobRetryHeap, 0, 16)
+	return ret
+}
+
+func (q *regionJobRetryer) run(ctx context.Context) {
+	for {
+		var front *regionJob
+		q.protectedQueue.mu.Lock()
+		if len(q.protectedQueue.q) > 0 {
+			front = q.protectedQueue.q[0]
+		}
+		q.protectedQueue.mu.Unlock()
+
+		switch {
+		case front != nil:
+			select {
+			case <-ctx.Done():
+				return
+			case <-q.done:
+				return
+			case <-q.reload:
+			case <-time.After(time.Until(front.waitUntil)):
+				q.protectedQueue.mu.Lock()
+				q.protectedToPutBack.mu.Lock()
+				q.protectedToPutBack.toPutBack = heap.Pop(&q.protectedQueue.q).(*regionJob)
+				// release the lock of queue to avoid blocking regionJobRetryer.push
+				q.protectedQueue.mu.Unlock()
+
+				// hold the lock of toPutBack to make sending to putBackCh and
+				// resetting toPutBack atomic w.r.t. regionJobRetryer.close
+				select {
+				case <-ctx.Done():
+					q.protectedToPutBack.mu.Unlock()
+					return
+				case <-q.done:
+					q.protectedToPutBack.mu.Unlock()
+					return
+				case q.putBackCh <- q.protectedToPutBack.toPutBack:
+					q.protectedToPutBack.toPutBack = nil
+					q.protectedToPutBack.mu.Unlock()
+				}
+			}
+		default:
+			// len(q.q) == 0
+			select {
+			case <-ctx.Done():
+				return
+			case <-q.done:
+				return
+			case <-q.reload:
+			}
+		}
+	}
+}
+
+// push should not be blocked for long time in any cases.
+func (q *regionJobRetryer) push(job *regionJob) bool {
+	q.protectedClosed.mu.Lock()
+	defer q.protectedClosed.mu.Unlock()
+	if q.protectedClosed.closed {
+		return false
+	}
+
+	q.protectedQueue.mu.Lock()
+	heap.Push(&q.protectedQueue.q, job)
+	q.protectedQueue.mu.Unlock()
+
+	select {
+	case q.reload <- struct{}{}:
+	default:
+	}
+	return true
+}
+
+// close will return the number of jobs that are not put back when first called.
+// close MUST be called when caller is sure that regionJobRetryer holds no job
+// or regionJobRetryer.run has exited.
```

Comment:
If `regionJobRetryer.run` has exited, what is `q.done` is used for?

Reply:
the other case is all job is finished, caller use `retryer.close()` to notify `retryer.run()`

Reply:
The semantics is so confusing. I suggest using `ctx` to notify `run` to exit.